### ISSUE TITLE
BUG: Python operator 'in' behaves in two different ways on Series object

### ISSUE THREAD

**Issue by gritschel** [pydata/pandas#14235]
The python operator 'in' acting on a pandas.Series object behaves in two different ways.
When iterating over the series it acts on the list of values. But for membership testing it acts on the index. Is this the desired behavior?

```
import pandas as pd
s = pd.Series([1, 2, 3])
for v in s: print v
    # prints 1 2 3, i.e. 'in' is acting on the values
for v in [1, 2, 3]: print v in s
    # prints True True False, i.e. the latter 'in' is acting on the index
```

This behavior also leads to peculiar behavior when working with DataFrames.

```
df = pd.DataFrame({'a': [1, 2, 3], 'b': [4, 5, 6]})
df

#      a  b
#  0  1  4
#  1  2  5
#  2  3  6

for value in df['a']:
    print value,
    if value in df['b']:
        print 'also in column b'

#  1 also in column b
#  2 also in column b
#  3
```

This is not what one would naively expect from that code snippet and it can potentially lead to errors that might be difficult to spot.
This is the behavior that is also mentioned in the docs (in a pretty hidden way -- one has to actively search for it), but I suggest that this should not be the desired behavior. Instead, I'd find it much more natural, if 'in' worked always on the values and never on the index. If needed one can always use 'in s.index', but the standard way should be acting on the values always.

pandas version 0.18.1

**Comment by jorisvandenbossche** [comment]
I agree it is confusing, but as you said, it is the documented behaviour (for iteration: http://pandas.pydata.org/pandas-docs/stable/basics.html#iteration, membership test: http://pandas.pydata.org/pandas-docs/stable/gotchas.html#using-the-in-operator although the argument that series is dict-like so it checks the keys/index and not the values, is not applied in the iteration case ...). 
Changing this will break users code, so is probably not going to happen soon (we could consider it in 2.0, but still it is not clear what would be the best solution: both checking values, or both checking index). But if you have suggestions to improve clarity in the docs, certainly welcome!

**Comment by cobalamin** [comment]
I agree with the OP. There are a few StackOverflow questions about this already (e.g., [here](https://stackoverflow.com/questions/49393053/using-in-operator-with-pandas-series/49393472#49393472), and [here](https://stackoverflow.com/questions/47189831/pandas-series-and-the-in-operator) from the opposite expectation perspective).

Let me make a few points:

- `x in y`, for objects `y` that are iterable, is generally expected to return `True` exactly if `for z in y: ...` sets `z` to an object equal to `y` at least once. [This is the default behaviour](https://docs.python.org/3/reference/expressions.html#membership-test-operations) of `in` when you  only implement `__iter__`, and all other Python libraries I've ever encountered adhere to this convention.
- Since it's claimed that Series are dict-like: `dict`s are consistent, both the `in` keyword and the `in` operator referring to keys; `Series` are not (you've pointed this out @jorisvandenbossche):
    ```python
    > dict = {'a': 3, 'b': 2}
    > for x in dict:
    >   print(x)
    a
    b

    > 3 in dict
    False
  
    > 'a' in dict
    True
    ```
- This inconsistency is the main reason for confusion as evidenced by the two linked StackOverflow posts, the first expecting `x in y` to be true if the value `x` is in y, the second expecting `for x in y` to yield keys.
- While it _is_ documented, it is documented in a gotcha section, a place were users will generally not look for something as "fundamental" as the `in` operator.
- The `isin` function often presented as an alternative for checking whether a Series contains a value is neither memory- nor runtime-efficient nor nicely readable (the needle and haystack being 'inverted'), and it requires a list argument even if one only wants to check for a single value, so from an intuitive standpoint for this specific purpose, it "checks the wrong way around": `s.isin([2])` checks for each value in the Series if it is in the list `[2]`.

I think making a consistent choice for either values or keys would be most sensible, since I don't see how this could be put in the docs in an appropriate way: IMHO, the current place (in gotchas) is too hidden, but any other place (e.g. in the Series API) is out of place, for the simple reason that `in` is a Python keyword/operator, and not part of the API of Series objects. IMHO, breaking user code is fine when this breaking change is introduced in a few versions into the future while affected code prints deprecation warnings (by printing warnings in the `__contains__` or the `__iter__` implementations).

**Comment by ParitoshSingh07** [comment]
Just got bit by this. It was miserable. agreed that it needs to be consistent, and imho `in` on the actual values seems much more sensible. iteration on series works on values, and no one really wants to check indices "implicitly".

**Comment by MarcoGorelli** [comment]
I agree that this is super-intuitive. Hardly anyone expects
```python
3 in pd.Series([1, 2, 3])
```
to return `False`.

I have a suggestion: deprecate `Series.__contains__`.

Concretely:

2.x behaviour:
```python
In [70]: 3 in pd.Series([1, 2, 3])
<ipython-input-70-1e925d619135>:1: FutureWarning: membership in Series is deprecated. Please use `value in series.index` to preserve the current behaviour, or `value in series.array` to check membership using the Series' values
Out[70]: False
```

3.0 behaviour:
```python
In [70]: 3 in pd.Series([1, 2, 3])
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
Cell In[70], line 1
----> 1 pd.Index.is_floating()

NotImplementedError: using `in` to check membership in Series is no longer allowed. Please use `value in series.index` to check for memership using the index, or `value in series.array` to check membership using the values
```

**Comment by MarcoGorelli** [comment]
time for another @pandas-dev/pandas-core @pandas-dev/pandas-triage tag, sorry y'all

**Comment by jbrockmendel** [comment]
AFAICT the Series behavior is incorrect bc it uses `IndexOpsMixin.__iter__` where it is "supposed to" use `NDFrame.__iter__`.  That said, i agree with Marco that deprecating `__contains__` is changing either.  (id be open to deprecating `__iter__` too)

**Comment by rhshadrach** [comment]
Both `__iter__` and `__contains__` for DataFrames using the columns. It therefore seems most natural that they use the index on a Series. This would also agree with a dictionary-analogy of Series. Rather than deprecating, I would suggest changing the `__iter__` to instead iterate over the index.

Both Series and DataFrames are data structures with a well-defined order, and as such I think users expect to be able to iterate over them. While there is ambiguity because there are multiple choices for what iteration / containment could mean, I think choosing a consistent definition is possible and helpful. I'm -0.5 on deprecating the dunders entirely.

**Comment by MarcoGorelli** [comment]
thanks for reading

>  I would suggest changing the __iter__ to instead iterate over the index.

are you sure this wouldn't be too disruptive? I think it would change
```python
In [3]: list(pd.Series([1,2,3]))
Out[3]: [0, 1, 2]
```
which seems like a really big change

It's unfortunate that when dealing with magic dunder methods, it's very difficult to do deprecations. We can hardly issue a warning like "the behaviour of `__iter__` is doing to change, please pass argument `x` to silence this warning and opt in to the new behaviour"

**Comment by rhshadrach** [comment]
> are you sure this wouldn't be too disruptive?

It would be no more disruptive than deprecating the method entirely, no?

> It's unfortunate that when dealing with magic dunder methods, it's very difficult to do deprecations.

This can be done using options like `future.series_iteraiton` that @jbrockmendel has proposed.

There is still an inconsistency however. I think users would expect that `list(ser)` and `ser.tolist()` behave the same, but I do feel `ser.tolist()` should return the values for reasons I haven't yet been able to express.

**Comment by jreback** [comment]
there have been lots of discussion about this over the years 

imho either leave this (and fix iteration) or deprecate entirely

changing to mean values is a non starter

**Comment by MarcoGorelli** [comment]
> It would be no more disruptive than deprecating the method entirely, no?

My thinking is that a loud error telling you what to use instead is more user-friendly than a sudden behaviour change (even if preceded by futurewarnings, which unfortunately people often either silence or don't read)

**Comment by rhshadrach** [comment]
> My thinking is that a loud error telling you what to use instead is more user-friendly than a sudden behaviour change (even if preceded by futurewarnings, which unfortunately people often either silence or don't read)

This is definitely true. But if we were to agree that changing `__iter__` is better for users (and I understand that's a big if), then I don't think we should impoverish the users who follow good practices to help those that don't.

**Comment by MarcoGorelli** [comment]
Sure, agree in general, it's the "changing `__iter__` being better for users" that I'm really questioning - I would also expect `ser.tolist()` and `list(ser)` to give the same result

**Comment by bashtage** [comment]
I think whether these return the same thing depends on whether `Series` should be more-dict like. 

I think it would be hard to change the behavior of `__iter__` without

a. Annoying users since they would need to change their code from `for val in ser` to be `for val in ser.array` anyway to silence the warning.
b. Not creating surprising results from code that hansn't been updated.

If this were an election, I'd vote for leaving `__contains__` and deprecating with eventual raise on `__iter__`.

**Comment by MarcoGorelli** [comment]
> If this were an election, I'd vote for leaving __contains__ and deprecating with eventual raise on __iter__.

Isn't `__contains__` the more counter-intuitive one?

Pandas developers may think of Series as dicts, but do users? Does the analogy even hold, given that Series can have duplicate row labels but dict keys must be unique?

Looks like this may be a topic for PDEP-13, there's more variety in opinion here than I was expecting - thanks all for weighing in!

**Comment by jbrockmendel** [comment]
xref #43014

**Comment by rhshadrach** [comment]
> Pandas developers may think of Series as dicts, but do users? 

For those who don't care about the index (likely always using a RangeIndex), most likely not. Those that do use meaningful indices I would assume so. But I have no idea what that split may look like.

> Does the analogy even hold, given that Series can have duplicate row labels but dict keys must be unique?

I don't think uniqueness destroys the dict analogy - especially regarding iteration and containment. But perhaps there are cases that do?

**Comment by kaiwu-astro** [comment]
I am a scientist in astronomy and this just bitted me. I feel lucky I discovered this before I publish my journal paper so that I did not publish wrong results in the frontier of science. But, are others aware of this? I feel worried. I think this is definitely a nightmare for scientists, and even a hidden bomb in science that could explode at any moment.

Pandas and Numpy are the most important tools for researchers. Since ```a in List``` and ```a in numpy.array``` behaves the same, and Pandas Series or DataFrame can be built based on list or numpy.array, I think most people will believe ```a in series``` means the same thing with no doubt. What's worse, it raises no error, and the (wrong) result is hidden below huge datasets which we usually do with DataFrame or its column Series. As a result, no one will be aware of this.

I know this is a documented behavior and you may have many reasons why ```a in series``` means ```a in series.index```, and of course you can blame users for not reading the doc. BUT, some of you may have heard about this: https://www.nature.com/articles/d41586-021-02211-4 The auto-correction in MS Excel has led to many unnoticed errors in research. That is the consequence. Microsoft is not wrong, but science is harmed. 

**Considering the leading role of Pandas in all subjects worldwide, I urge the developers think about this carefully. A simple implementation of ```a in series```=```a in series.values``` would save loads of errors in the future.**

(I asked several colleagues and none of them knew this before, and we rushed to check all our Python pipelines.... but did other millions of researchers know this? I am worried...)

**Comment by Dr-Irv** [comment]
> **Considering the leading role of Pandas in all subjects worldwide, I urge the developers think about this carefully. A simple implementation of `a in series`=`a in series.values` would save loads of errors in the future.**

This is a debate.  Is a `Series` like a `dict` or like an array?  It's always been thought of as a `dict`, so the behavior is like a `dict`:
```python
>>> import pandas as pd
>>> d={1:"a", 2:"b", 3:"c"}
>>> 1 in d
True
>>> "a" in d
False
>>> s = pd.Series(d)
>>> s
1    a
2    b
3    c
dtype: object
>>> 1 in s
True
>>> "a" in s
False
```

**Comment by rhshadrach** [comment]
Can there be bad behaviors that result from having a container that appears to be a sequence but raises on calls to `__contains__`? If this is okay, I can be on board with deprecating.

I do think we should make a decision on `__iter__` as well before deprecating.
